server/models/meshery_controllers.go

```diff
@@ -36,146 +36,157 @@ const (
+	ctxMeshsyncDataHandler *MeshsyncDataHandler
 
 	log          logger.Handler
 	oprDepConfig controllers.OperatorDeploymentConfig
 	dbHandler    *database.Handler
 }
 
-func (mch *MesheryControllersHelper) GetControllerHandlersForEachContext() map[string]map[MesheryController]controllers.IMesheryController {
-	return mch.ctxControllerHandlersMap
+func (mch *MesheryControllersHelper) GetControllerHandlersForEachContext() map[MesheryController]controllers.IMesheryController {
+	return mch.ctxControllerHandlers
 }
 
-func (mch *MesheryControllersHelper) GetMeshSyncDataHandlersForEachContext() map[string]MeshsyncDataHandler {
-	return mch.ctxMeshsyncDataHandlerMap
+func (mch *MesheryControllersHelper) GetMeshSyncDataHandlersForEachContext() *MeshsyncDataHandler {
+	return mch.ctxMeshsyncDataHandler
 }
 
-func (mch *MesheryControllersHelper) GetOperatorsStatusMap() map[string]controllers.MesheryControllerStatus {
-	return mch.ctxOperatorStatusMap
+func (mch *MesheryControllersHelper) GetOperatorsStatusMap() controllers.MesheryControllerStatus {
+	return mch.ctxOperatorStatus
 }
 
 func NewMesheryControllersHelper(log logger.Handler, operatorDepConfig controllers.OperatorDeploymentConfig, dbHandler *database.Handler) *MesheryControllersHelper {
 	return &MesheryControllersHelper{
-		ctxControllerHandlersMap:  make(map[string]map[MesheryController]controllers.IMesheryController),
-		log:                       log,
-		oprDepConfig:              operatorDepConfig,
-		ctxOperatorStatusMap:      make(map[string]controllers.MesheryControllerStatus),
-		ctxMeshsyncDataHandlerMap: make(map[string]MeshsyncDataHandler),
-		dbHandler:                 dbHandler,
+		ctxControllerHandlers:  make(map[MesheryController]controllers.IMesheryController),
+		log:                    log,
+		oprDepConfig:           operatorDepConfig,
+		ctxOperatorStatus:      controllers.Unknown,
+		// The nil check is performed for the ctxMeshsyncDataHandler and if it is nil, then a new dataHandler for the context is assigned.
+		// The presence of a handler for a context in a map indicate that the meshsync data for that context is properly being handled.
+		// Resetting this value results in again subscribing to the Broker.
+		ctxMeshsyncDataHandler: nil,
+		dbHandler:              dbHandler,
 	}
 }
 
 // initializes Meshsync data handler for the contexts for whom it has not been
 // initialized yet. Apart from updating the map, it also runs the handler after
 // updating the map. The presence of a handler for a context in a map indicate that
 // the meshsync data for that context is properly being handled
-func (mch *MesheryControllersHelper) UpdateMeshsynDataHandlers(ctx context.Context, connectionID, userID, mesheryInstanceID uuid.UUID, provider Provider) *MesheryControllersHelper {
+func (mch *MesheryControllersHelper) AddMeshsynDataHandlers(ctx context.Context, k8scontext K8sContext, userID, mesheryInstanceID uuid.UUID, provider Provider) *MesheryControllersHelper {
```

Comment:
How are we preparing for the support of Environments?

Reply:
Enquiring minds want to know. :)

Reply:
Because environment resources are logical groups of resources that server primary concerns with, hence when the user is issued a session/environment is changed from the switcher the necessary environment related details will be send to the UI and with each request envID needs to be included.

All APIs in the server will be updated to include envID.  As a result you should expect the above handler (and similarly other) have a envID param passed in their context (`ctx`).
In some cases, connectionID is enough (as it itself is an primary key) to manage connection, but including envID adds a direction.

Reply:
I'll request that we draw a hard line on whether Environment ID becomes a required parameter or not and suggest that we lean into the former.